### ISSUE TITLE
[Critical] - BatchDirectory is serialized and sent to client

### ISSUE THREAD

**Issue by gentledepp** [Mimetis/Dotmim.Sync#114]
When doing our first checks with dotmim.sync, we used a Xamarin.Android based client.
To our suprise, the agents ".Configuration.BatchDirectory" was set to the apps local folder (just as it should be, as it does not have write access anywhere else).
But during the sync process, we got some "UnauthorizedAccessExceptions" telling us that the temp folder of the server machine was inaccessible.

The reason is, that the SyncConfiguration.BatchDirectory is actually serialized and sent to the client.
So during the sync process, the agents batchdirectory is updated with the one from the server, which makes synchronization impossible.

Frankly, I do not see the reason why you would **ever** want to send the batchdirectory to the client. It is the same as sending the connectionstring to the client. Isn't it?

**Comment by lordofffarm** [comment]
Hi, this has been fixed here #108

**Comment by gentledepp** [comment]
This issue is not resolved at all.
I just debugged it and found out, that the BatchDirectory, which is configured on the server side, is repeatedly overwritten with client configuration:
```

    private async Task<HttpMessage> BeginSessionAsync(HttpMessage httpMessage)
        {
            HttpMessageBeginSession httpMessageBeginSession;
            if (httpMessage.Content is HttpMessageBeginSession)
                httpMessageBeginSession = httpMessage.Content as HttpMessageBeginSession;
            else
                httpMessageBeginSession = (httpMessage.Content as JObject).ToObject<HttpMessageBeginSession>();

            // CHANGE: use configuration.batchdirectory and just fallback to httpmessagebeginsession...
            var batchDirectory = this.Configuration?.BatchDirectory ?? httpMessageBeginSession.SyncConfiguration.BatchDirectory;
            
            // the Conf is hosted by the server ? if not, get the client configuration
            httpMessageBeginSession.SyncConfiguration = 
                this.Configuration ?? httpMessageBeginSession.SyncConfiguration;

            // Begin the session, requesting the server for the correct configuration
            (SyncContext ctx, SyncConfiguration conf) =
                await this.BeginSessionAsync(httpMessage.SyncContext, 
                        httpMessageBeginSession as MessageBeginSession);

            // One exception : don't touch the batch directory, it's very client specific and may differ from server side
            conf.BatchDirectory = batchDirectory;
            httpMessageBeginSession.SyncConfiguration = conf;

            httpMessage.SyncContext = ctx;

            // Dont forget to re-assign since it's a JObject, until now
            httpMessage.Content = httpMessageBeginSession;

            return httpMessage;
        }


```

and:
```


        public async Task<(SyncContext, SyncConfiguration)> BeginSessionAsync(SyncContext ctx, MessageBeginSession message)
        {
            // CHANGE: Manually override batchdirectory one very request to resolve this issue!!!
            if (this.Configuration.BatchDirectory != null)
                message.SyncConfiguration.BatchDirectory = this.Configuration.BatchDirectory;

            return await this.LocalProvider.BeginSessionAsync(ctx, message);
        } 
        public async Task<(SyncContext, List<ScopeInfo>)> EnsureScopesAsync(SyncContext ctx, MessageEnsureScopes message)
            => await this.LocalProvider.EnsureScopesAsync(ctx, message);
        public async Task<(SyncContext, DmSet)> EnsureSchemaAsync(SyncContext ctx, MessageEnsureSchema message)
            => await this.LocalProvider.EnsureSchemaAsync(ctx, message);
        public async Task<SyncContext> EnsureDatabaseAsync(SyncContext ctx, MessageEnsureDatabase message)
            => await this.LocalProvider.EnsureDatabaseAsync(ctx, message);

        public async Task<(SyncContext, BatchInfo, ChangesSelected)> GetChangeBatchAsync(SyncContext ctx,
            MessageGetChangesBatch message)
        {
            // CHANGE: Manually override batchdirectory one very request to resolve this issue!!!
            if (this.Configuration.BatchDirectory != null)
                message.BatchDirectory = this.Configuration.BatchDirectory;
            
            return await this.LocalProvider.GetChangeBatchAsync(ctx, message);
        }

        public async Task<(SyncContext, ChangesApplied)> ApplyChangesAsync(SyncContext ctx, MessageApplyChanges message)
            => await this.LocalProvider.ApplyChangesAsync(ctx, message);
        public async Task<(SyncContext, Int64)> GetLocalTimestampAsync(SyncContext ctx, MessageTimestamp message)
            => await this.LocalProvider.GetLocalTimestampAsync(ctx, message);
        public async Task<SyncContext> WriteScopesAsync(SyncContext ctx, MessageWriteScopes message)
            => await this.LocalProvider.WriteScopesAsync(ctx, message);
        public async Task<SyncContext> EndSessionAsync(SyncContext ctx)
            => await this.LocalProvider.EndSessionAsync(ctx);

```

had to be added to fix this problem.

Honestly, I do not know if this is completely correct.

Also, I cannot provide a PR as long as the master branch has most of the tests just commented out. We'll continue on our repo and potentially diverge from this repository eventually
